### ISSUE TITLE
gskcapicmd_64 100% cpu usage and freeze start mq container

### ISSUE THREAD

**Issue by kalekhin** [ibm-messaging/mq-container#462]
Hello! 

When my laptop is connected to a power source, the gskcapicmd_64 call lasts forever, and one thread uses 100% of the processor. This is fixed if the laptop is powered by a battery. I don't understand why and how this is related.

Reproduced on ROG Zephyrus G15 GA503
GA503QM-HN094 (AMD Ryzen 7 5800HS with Radeon).
Env: Win 10 Pro 20 H2 build 20 H2 with all updates for 17/07/2021
Docker toolbox v19. 03. 1 is running on a virtual box 6.1.22 r144080 (Qt5.6.2) with the extension package 6.1.22 r144080.

Just run:
docker run \
--env LICENSE=accept \
--env MQ_QMGR_NAME=QM1 \
--publish 1414:1414 \
--publish 9443:9443 \
--detach \
ibmcom/mq

And the bug will happen.
![image](https://user-images.githubusercontent.com/55395720/126017369-00e38eb1-b70e-4692-b4d3-aab645e152e4.png)

https://github.com/ibm-messaging/mq-container/blob/4580cecf4973107dff184e8cbbcf9ac7f5b4e7df/internal/keystore/keystore.go#L192

**Comment by kalekhin** [comment]
The problem is solved if I provide all the processor cores for the linux virtual machine where docker is installed, or if I setup the paravirtualization interface for this machine as hyper-v.

In the first solution, the CPU load is 100% already by the java process in the mq container. But restarting the container has a chance to solve the problem (only a chance).
The second solution looks more stable. So far, there have been no problems

I am not sure that any solution is stable.

**Comment by jfmatheusg** [comment]
![image](https://user-images.githubusercontent.com/33701513/132759946-9d41413f-6f7c-486b-b7da-97bf444826e5.png)

Same problem here :(
Im running the container on Ubuntu 20.04 LTS, AMD Ryzen 7 5800, Lenovo Legion 5 Pro.

**Comment by mihmig** [comment]
I have a similar situation - 100% CPU load by gsk8capicmd_64:
![image](https://user-images.githubusercontent.com/4863752/136001033-f84f0ed5-7684-44c0-b345-06908642577c.png)
my hardware/software set::
AMD Ryzen 7 5800H
Windows 10.0.19042.1237 with WSL 2 with core version 5.10.16
Docker 20.10.8 build 3967b7d

**Comment by andreysaksonov** [comment]
AMD Ryzen 7 5800H (Lenovo Legion 5)
Fedora 36 (kernel 5.17.12-300.fc36.x86_64), Docker Desktop 4.9.0 (docker 20.10.16)

same issue: `/opt/mqm/gskit8/bin/gsk8capicmd_64 -keydb -create -type cms -db /run/runmqserver/tls/key.kdb -pw JtM27EP7L2LH -stash` loads CPU 100% and takes from 6 to 40 minutes (randomly)

![image](https://user-images.githubusercontent.com/2135309/172321886-52cd63d7-c14d-43e0-9abd-9a4c7862caf5.png)

**Comment by parrobe** [comment]
Hi @andreysaksonov , @mihmig , @jfmatheusg , @kalekhin .

Arthur has asked me to look into this. The gsk8capicmd is owned by a separate internal IBM team to IBM MQ but i can raise a support ticket with them to ask them to take a look. To help them diagnose the issue they are likely to want trace of the issue. 

Please could you run the same commands as before that caused the 100% CPU issue with the `-trace <file>` option. For example, taking @andreysaksonov 's command i would run: `/opt/mqm/gskit8/bin/gsk8capicmd_64 -keydb -create -type cms -db /run/runmqserver/tls/key.kdb -pw JtM27EP7L2LH -stash -trace /tmp/trace.output`.

Please then send me the file generated; in the previous example i would need `/tmp/trace.output`.
You can send the file by either attaching it to a comment here or directly via email to `parrobe@uk.ibm.com`.

In the meantime I'll get the ball rolling with GSKit and hopefully we can get to the bottom of this.

**Comment by parrobe** [comment]
Thanks @andreysaksonov - I've passed these details onto the GSKit team. I will let you know when they have responded.

**Comment by andreysaksonov** [comment]
Yes, it solves the issue, thanks. I will leave link to original GSKit bug: https://www.ibm.com/support/pages/apar/IJ28497

**Comment by andyedwardsibm** [comment]
We've started hitting this in App Connect in containers too.

When considering the comment above from `parrobe`, (that basically says "you must set this to fix the problem, but must not set it if you don't see the problem) we've identified an interesting problem when running in a cluster.  Initially, one might think that it's safe to set `ICC_SHIFT` on a cluster, and so we could add a config parameter for an operand so that users can selectively set `ICC_SHIFT`

But the cluster may have a mix of hardware for the workers.  Some may use AMD chips susceptible to this problem, some may use Intel chips.  This would mean that it depends what worker a Pod lands on as to whether the user should set the config parameter or not, and that Pod placement is, by default, random

**Comment by imavo** [comment]
I understand that there has been an update (fixpack) to the IBM GSKit component that addresses this cpu specific-issue, so you could contact IBM support to get the updated GSKit and apply it to your estate, or instead  wait for the next fixpack update to you rmq component(s) to include the updated GSKit . Hopefully that would then mean a permanent fix without needing workarounds like ICC_SHIFT=3

**Comment by parrobe** [comment]
Hi @imavo - IBM MQ aims to update all our thirdparty component versions to the latest with each released version of IBM MQ. Unfortunately, there are delays taking the latest GSKit version due to issues found through our regression testing. We will update as soon as we are able.
